Implement #359 - Force type tags in the PReader/PBuilder API.

jsuereth opened the pull request:
Beacuse type tags are a fundamental aspect of the design of Pickler/Unpicklers,
we now FORCE their existence for every type.
- `beginEntry` on PBuilder now takes a `FastTypeTag[_]`
- `hintStaticallyElided`, `hintDynmaicallyElided` and `hintTag` have all
  been unified into a `hintElidedType` which takes the elided tag.
- Fix both JSON + Binary format to handle the new design
- Generally simplifies hand-written picklers
- Fixed an issue with `Null` handling in JSON format.

Fixes #359 
Review by @phaller  and @eed3si9n

jsuereth committed:
Implement #359 - Force type tags in the PReader/PBuilder API.

Beacuse type tags are a fundamental aspect of the design of Pickler/Unpicklers,
we now FORCE their existence for every type.

- `beginEntry` on PBuilder now takes a `FastTypeTag[_]`
- `hintStaticallyElided`, `hintDynmaicallyElided` and `hintTag` have all
  been unified into a `hintElidedType` which takes the elided tag.
- Fix both JSON + Binary format to handle the new design
- Generally simplifies hand-written picklers
- Fixed an issue with `Null` handling in JSON format.

jsuereth committed:
Fix test failure check for travis.

phaller commented on core/src/test/scala/pickling/run/transient2.scala:
```diff
@@ -75,9 +75,10 @@ class Transient2SparkTest extends FunSuite {
+    //implicit def depP = PicklerUnpickler.generate[Dependency[_]]
+    //implicit val seqDepp = Defaults.seqPickler[Dependency[_]]
+    //implicit val rddpp = PicklerUnpickler.generate[RDD[Int]]
+    //implicit val fmrddp = PicklerUnpickler.generate[FlatMappedRDD[Int, Int]]
```
Good to know these are no longer needed.

phaller commented on core/src/test/scala/pickling/run/non-public-separate.scala:
```diff
@@ -7,7 +7,7 @@ class Person(private val name: String, age: Int, val hobby: Hobby) {
   override def toString = s"Person(name = $name, hobby = $hobby)"
 }
-class Hobby(var name: String, private var notes: String, private val attitude: String) {
+final class Hobby(var name: String, private var notes: String, private val attitude: String) {
```
What's the reason? Apparently, the test was passing also without `final`?

jsuereth replied:
good question, I'm not sure why I did that.  Perhaps to test out the elide?

jsuereth replied:
OOOH, so this is actually an `isEffectivelyFinal` bug.  It appears to return true for the pickler and not true for the unpickler....

phaller commented:
Nice to see the simplifications in the hand-written picklers. Also: I like the disappearing `hintTag` calls that used to be sprinkled around the macro code.

phaller commented:
LGTM

jsuereth committed:
Fixes and issue with dynamic-elides so that we have uniformity.

- We no longer mark elided type in dynamic-elide situations
- Unpicklers do not have to support the ontion of dynamic elide.

jsuereth commented:
@phaller Ok, so the test change was due to dynamic elides no longer working.  That's, in part, because you cannot mirror the dynamic elide logic at the unpickler side, so I've disabled them until we can figure out how to fully mirror the behavior on both sides of the pickler/unpickler boundary (until then it's dubious logic).
